Implement spectator list display

bdach opened the pull request:
- First step for https://github.com/ppy/osu/issues/22087
- Supersedes / closes https://github.com/ppy/osu/pull/22795

https://github.com/user-attachments/assets/797eacb6-0315-4568-9f87-e969e4f542c9

Roughly uses design shown in https://github.com/ppy/osu/pull/22795#issuecomment-1579936284 with some modifications to better fit everything else, and some customisation options so it can fit better on other skins.

bdach committed:
Implement spectator list display

- First step for https://github.com/ppy/osu/issues/22087
- Supersedes / closes https://github.com/ppy/osu/pull/22795

Roughly uses design shown in
https://github.com/ppy/osu/pull/22795#issuecomment-1579936284 with some
modifications to better fit everything else, and some customisation
options so it can fit better on other skins.

peppy committed:
Fix idle/hover colour handling weirdness in `OsuHoverContainer`

peppy committed:
Adjust test to work better when running in sequence

peppy committed:
Avoid list width changing when spectator count changes

peppy committed:
Add gradient on final spectator if more than list capacity are displayed

peppy committed:
Adjust animations a bit

Removed autosize duration stuff because it looks weird when the list is
shown from scratch where users are already fully populated in it.

peppy commented:
I pushed test fixes and some visual changes. @bdach please confirm you're okay with the new commits.

bdach committed:
Merge branch 'master' into spectator-list-visuals

bdach committed:
Fix potential index accounting mistake when creating spectator list with spectators already present

Noticed by accident, but if the `BindCollectionChanged()` callback fires
immediately in `LoadComplete()` when set up and there are spectators
present already, then `NewStartingIndex` in the related event is -1:

	https://github.com/dotnet/runtime/blob/b03f83de362f7168c94daa2f4b192959abefe366/src/libraries/System.ObjectModel/src/System/Collections/Specialized/NotifyCollectionChangedEventArgs.cs#L84-L92

which kinda breaks the math introducing off-by-ones and in result causes
11 items to be displayed together rather than 10.
